subprojects/core/src/main/java/org/gradle/api/internal/file/copy/DuplicateHandlingCopyActionDecorator.java

```diff
@@ -23,49 +23,59 @@
 import org.gradle.api.internal.DocumentationRegistry;
 import org.gradle.api.tasks.Copy;
 import org.gradle.api.tasks.WorkResult;
+import org.gradle.util.internal.TextUtil;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import java.util.HashSet;
-import java.util.Set;
+import java.io.File;
+import java.util.HashMap;
+import java.util.Map;
 
 public class DuplicateHandlingCopyActionDecorator implements CopyAction {
 
     private final static Logger LOGGER = LoggerFactory.getLogger(DuplicateHandlingCopyActionDecorator.class);
     private final CopyAction delegate;
+    private final CopySpecInternal spec;
     private final DocumentationRegistry documentationRegistry;
 
-    public DuplicateHandlingCopyActionDecorator(CopyAction delegate, DocumentationRegistry documentationRegistry) {
+    public DuplicateHandlingCopyActionDecorator(CopyAction delegate, CopySpecInternal spec, DocumentationRegistry documentationRegistry) {
         this.delegate = delegate;
+        this.spec = spec;
         this.documentationRegistry = documentationRegistry;
     }
 
     @Override
     public WorkResult execute(final CopyActionProcessingStream stream) {
-        final Set<RelativePath> visitedFiles = new HashSet<>();
+        final Map<RelativePath, String> visitedFiles = new HashMap<>();
 
-        return delegate.execute(action -> stream.process(details -> {
+        return delegate.execute(action -> stream.process((FileCopyDetailsInternal details) -> {
```

Comment:
🤔 maybe we can just make FileCopyDetails a Describable too?

Reply:
We can (and I will) do that, but we still need the internal type here to use `isDefaultDuplicatesStrategy().`